instr/tests/src/test_instr_decoding.cpp

```diff
@@ -0,0 +1,85 @@
+#include <gtest/gtest.h>
+
+#include <sim/common.hpp>
+#include <sim/instr.hpp>
+#include <iostream>
+
+namespace sim {
+namespace instr {
+
+
+TEST(Instr, undef)
+{
+    Instr test = Instr(static_cast<InstrCode>(0));
+    ASSERT_EQ(test.id(), InstrId::UNDEF);
+}
+
+TEST(Instr, add)
+{
+    Instr test = Instr(static_cast<InstrCode>(0x00710a13)); //ADDI x20, x2, 7
+    ASSERT_EQ(test.id(), InstrId::ADDI);
+    ASSERT_EQ(test.rd(), 20);
+    ASSERT_EQ(test.rs1(), 2);
+    ASSERT_EQ(test.imm(), 7);
+}
+
+TEST(Instr, srli)
+{
+    Instr test = Instr(static_cast<InstrCode>(0x02175113)); //srli x2, x14, 33
+    ASSERT_EQ(test.id(), InstrId::SRLI);
+    ASSERT_EQ(test.rd(), 2);
+    ASSERT_EQ(test.rs1(), 14);
+    ASSERT_EQ(test.imm(), 33);
+}
+
+TEST(Instr, jal)
+{
+    Instr test = Instr(static_cast<InstrCode>(0x02200a6f)); //jal x20, 34
+    ASSERT_EQ(test.id(), InstrId::JAL);
+    ASSERT_EQ(test.rd(), 20);
+    ASSERT_EQ(test.imm(), 34);
+}
+
+TEST(Instr, garbage)
+{
+    Instr test = Instr(static_cast<InstrCode>(0xfffa056f)); //hz, absolutno
```

Comment:
Please, improve this test:
Iterate over all uint32_t values. Program must end successfully

Reply:
added such test, takes long time though